csharp/src/Ice/AnyClass.cs

```diff
@@ -39,24 +30,16 @@ public virtual void iceWrite(OutputStream ostr)
             ostr.EndClass();
         }
 
+        // Read all the fields of this instance from the stream. See InputStream.
         [EditorBrowsable(EditorBrowsableState.Never)]
-        public virtual void iceRead(InputStream istr)
-        {
-            istr.StartClass();
-            iceReadImpl(istr);
-            istr.EndClass(false);
-        }
+        protected abstract void IceRead(InputStream istr, bool firstSlice);
+        internal void Read(InputStream istr) => IceRead(istr, true);
```

Comment:
It's useful because IceRead is protected, and Read is internal.
I initially made IceRead "protected internal" but this has two downsides:
- all generated IceRead in the Ice assembly must be marked protected internal as well, while the generated IceRead outside the Ice assembly must be marked just protected, and slice2cs has no way to know whether a definition will end up in the Ice assembly or not
- there is no equivalent to protected internal in other language mappings

Here, having two methods (protected IceRead and internal Read) is the cleanest.
The internal Read is named Read and not IceRead because it's internal and therefore cannot conflict with user code (it can conflict with other code in the Ice assembly but that's not user code).

Reply:
What happens if the users class has a Read data member, we will have to declare it with `new int Read` to avoid warnings of hiding base method Read

Reply:
No because Read is internal.